gordon_janitor_gcp/reconciler.py

```diff
@@ -0,0 +1,243 @@
+        self.dns_client._session.close()
+
+        msg = ('Reconciliation of desired records against actually records in '
+               'Google DNS is complete.')
+        logging.info(msg)
+
+    async def publish_change_messages(self, desired_rrsets, action='additions'):
+        """Publish change messages to the ``changes_channel``.
+
+        NOTE: Only `'additions'` are currently supported. `'deletions'`
+        may be supported in the future.
+
+        Args:
+            desired_rrsets (list(GCPResourceRecordSet)): Desired record
+                sets that are not in Google Cloud DNS.
+            action (str): (optional) action for these corrective
+                messages. Defaults to ``'additions'``.
+        """
+        for rrset in desired_rrsets:
+            msg = {
+                'resourceRecords': attr.asdict(rrset),
+                'action': action
+            }
+            # TODO (lynn): add metrics.incr call here once aioshumway is
+            #              released
+            logging.debug('Creating the following change message: {msg}')
+            await self.changes_channel.put(msg)
+
+        logging.info(f'Created {len(desired_rrsets)} change messages.')
+
+    def _parse_rrset_message(self, message):
+        # assert that keys 'zone' and 'rrsets' are present, and return
+        # values for each
+        try:
+            zone = message['zone']
+        except KeyError:
+            msg = (f'No zone was defined in the given message: {message}.')
+            logging.error(msg)
+            raise exceptions.GCPGordonJanitorError(msg)
+        try:
+            rrsets = message['rrsets']
+        except KeyError:
+            msg = (f'No resource record sets were defined in given message: '
+                   f'{message}.')
+            logging.error(msg)
+            raise exceptions.GCPGordonJanitorError(msg)
+
+        return zone, rrsets
+
+    async def validate_rrsets_by_zone(self, zone, rrsets):
+        """Given a zone, validate current versus desired rrsets.
+
+        If there are any missing records, a corrective message for each
+        record will be published.
+
+        Args:
+            zone (str): zone to query Google Cloud DNS API.
+            rrsets (list): desired record sets to which to compare the
+                Cloud DNS API's response.
+        """
+        desired_rrsets = [
+            cloud_dns.GCPResourceRecordSet(**record) for record in rrsets
+        ]
+
+        actual_rrsets = await self.dns_client.get_records_for_zone(zone)
+
+        # NOTE: only working on "additions" (which includes changes to
+        #       current records) right now.
+        # TODO: (FEATURE) add support for cleaning up records that
+        #       should have been deleted.
+        missing_rrsets = [
+            rs for rs in desired_rrsets if rs not in actual_rrsets
+        ]
+        await asyncio.ensure_future(
```

Comment:
(metrics) This could be a nice spot to emit the count of missing_rrsets. Or accumulate it and flush the final number in `done`, perhaps.

Future idea: count additions and deletions separately.